apiserver/params/network.go

```diff
@@ -310,15 +313,33 @@ func InterfaceInfoFromNetworkConfig(configs []NetworkConfig) network.InterfaceIn
 			Origin:              network.Origin(v.NetworkOrigin),
 		}
 
-		// Compatibility layer for older clients that do not populate
-		// Addresses/ShadowAddresses.
-		if len(result[i].Addresses) == 0 && v.Address != "" {
-			result[i].Addresses = append(
-				result[i].Addresses,
-				network.NewProviderAddress(v.Address),
-			)
+		// Compatibility accommodations follow.
+		// TODO (manadart 2021-03-05): Juju 3/4 should require that only the
+		// address collections are used, and the following fields removed from
+		// the top-level interface:
+		// - CIDR
+		// - ConfigType
+		// - Address
+
+		// 1) For clients that populate Addresses, but still set
+		//    address-specific fields on the device.
+		//    Note that the assumption must hold (as it does at the time of
+		//    writing) that the collections are only populated with a single
+		//    member, with repeated devices for each address.
+		if len(result[i].Addresses) > 0 {
```

Comment:
In forthcoming patches, `ConfigType` will come off `InterfaceInfo` and be dealt with in this block as CIDR is.